### ISSUE TITLE
SugarCrepe Diagnostic Evaluation + COCO Failure Analysis

### ISSUE THREAD

**Issue by iremurhan** [iremurhan/tez_v2_clean#60]
## Motivation
 
We need to understand **why** COCO lags behind Flickr (R@1 ~66 vs ~91.5) and **which failure modes** our proposed methods (B4 classification, B5 segment embedding) should fix. SugarCrepe provides a controlled diagnostic: it tests compositional understanding broken down into specific sub-categories, without the benchmark artifacts that plague ARO.
 
### Part A: SugarCrepe Evaluation on All Checkpoints
 
Run the [SugarCrepe benchmark](https://github.com/RAIVNLab/sugar-crepe) on every finished checkpoint.
 
**SugarCrepe sub-categories:**
- **Replace** — object / attribute / relation words replaced
- **Swap** — word order swapped (changes relations)
- **Add** — extra objects/attributes added to caption
 
**Checkpoints to evaluate:**
- [x] B0 (all seeds, Flickr + COCO)
- [x] B0+ (all seeds)
- [x] B1 (all seeds, if COCO run recovers)
- [x] B2 (all seeds)
- [ ] B3 — LoRA (when ready)
- [ ] B4 — multi-label classification (when ready)
- [ ] B5 — segment-aware (when ready)
 
**Output:** Accuracy per sub-category per checkpoint, logged to WandB.
 
### Part B: COCO Failure Case Extraction
 
From the **B0 COCO checkpoint** (best baseline):
 
1. Run retrieval on COCO 5K test set.
2. Extract the **100 worst failures** (highest-rank errors) for both I2T and T2I.
3. For each failure, record:
   - Query (image or caption)
   - Ground-truth match
   - Model's top-1 retrieval
   - Rank of ground-truth in retrieved list
4. Manually categorize each failure into:
   - **Wrong object** — retrieved result has different main object
   - **Wrong relation** — correct objects but wrong spatial/action relation
   - **Wrong attribute** — correct object but wrong color/size/state
   - **Spatial confusion** — correct objects and relations but wrong arrangement
   - **Counting error** — wrong number of objects
   - **Other**
 
**Output:** A CSV/spreadsheet + a summary bar chart showing failure distribution.
 
### Part C: Cross-Analysis
 
- Correlate SugarCrepe sub-category weaknesses with COCO failure categories.
- If B0 is weak on SugarCrepe-Replace-Relation AND COCO failures cluster on "wrong relation" / "spatial confusion" → strong motivation for segment-aware approach.
 
## WandB Logging Strategy
 
**Goal:** Log SugarCrepe results to the **same WandB run** as the corresponding training experiment, so all metrics for a given run (training curves + retrieval metrics + SugarCrepe diagnostics) live in one place.
 
**Implementation approach:**
 
```python
import wandb
 
# Resume the existing WandB run for this checkpoint
# Each training run already has a run ID stored in the checkpoint or config
wandb.init(
    project="clip-finetune",
    id=existing_run_id,       # the WandB run ID from training
    resume="must",            # attach to existing run, don't create new
)
 
# Log SugarCrepe results as summary metrics (not step metrics)
# This adds them to the run's summary without interfering with training curves
wandb.run.summary["sugarcrepe/replace_object"] = 0.85
wandb.run.summary["sugarcrepe/replace_attribute"] = 0.78
wandb.run.summary["sugarcrepe/replace_relation"] = 0.62
wandb.run.summary["sugarcrepe/swap"] = 0.71
wandb.run.summary["sugarcrepe/add"] = 0.83
wandb.run.summary["sugarcrepe/overall"] = 0.76
 
# Optionally log the failure analysis as a WandB Table
failure_table = wandb.Table(
    columns=["query_type", "query", "gt_match", "top1_retrieval", "gt_rank", "failure_category"],
    data=failure_rows  # list of lists
)
wandb.log({"coco_failure_analysis": failure_table})
 
wandb.finish()
```
 
**Why `summary` instead of `log`?**
- `wandb.log()` creates a new step on the x-axis — it would appear as an extra training step, which is misleading.
- `wandb.run.summary` adds key-value pairs to the run's summary panel, visible in the runs table and comparable across runs. You can add columns like `sugarcrepe/replace_relation` to the WandB runs table and sort/filter by them.
 
**Alternative — if run IDs are lost or hard to recover:**
Create a separate eval run but link it via tags/groups:
```python
wandb.init(
    project="clip-finetune",
    name=f"sugarcrepe_eval_{run_name}",
    group=run_name,  # groups training + eval runs together
    job_type="eval",
    tags=["sugarcrepe", run_name],
)
```
 
## Deliverables
 
1. **Script:** `scripts/eval_sugarcrepe.py` — takes a checkpoint path, runs SugarCrepe, logs to WandB.
2. **Script:** `scripts/extract_failures.py` — takes a checkpoint path, runs COCO retrieval, extracts worst 100 failures as CSV.
3. **Results:** Summary table comparing all checkpoints across SugarCrepe sub-categories.
4. **Analysis doc:** Short write-up (can be in the thesis) correlating SugarCrepe weaknesses with COCO failure modes.
 
## Ordering
 
- Run **immediately after Issue 1 (B4)** implementation, but can start on B0/B0+/B1/B2 checkpoints right now since those are already finished (Flickr) or finishing (COCO).
- Re-run on B4 and B5 checkpoints as they become available.
